### PR TITLE
Rewrite the SbatLevel to use stream offsets and add unit tests

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: libfwupdplugin/fu-sbatlevel-section.c [hunk 1]
```diff
@@ -26,22 +28,21 @@ fu_sbatlevel_section_add_entry(FuFirmware *firmware,
...
-
-	/* look for the null terminator */
-	for (size = 0; ((offset + size) < bufsz); ++size) {
-		if (buf[offset + size] == 0)
+	for (guint i = offset; i < streamsz; i++) {
+		guint8 value = 0;
+		if (!fu_input_stream_read_u8(stream, i, &value, error))
+			return FALSE;
+		if (value == 0x0) {
+			streamsz = i - 1;
```
Comment (valentindavid) [inline libfwupdplugin/fu-sbatlevel-section.c:43]: I suppose we used to parse with the nul character included but now it does not take it into account. I suppose this is fine.

File: libfwupdplugin/fu-sbatlevel-section.c [hunk 2]
```diff
@@ -52,16 +53,15 @@ fu_sbatlevel_section_add_entry(FuFirmware *firmware,
...
 	fu_firmware_set_id(entry_fw, entry_name);
 	fu_firmware_set_offset(entry_fw, offset);
-	entry_blob = fu_bytes_new_offset(fw, offset, size, error);
-	if (entry_blob == NULL)
-		return FALSE;
-	if (!fu_firmware_add_image_full(firmware, entry_fw, error))
-		return FALSE;
-	if (!fu_firmware_parse(entry_fw, entry_blob, flags, error)) {
+	partial_stream = fu_partial_input_stream_new(stream, offset, streamsz - offset);
+	if (!fu_firmware_parse_stream(entry_fw, partial_stream, 0, flags, error)) {
```
Comment (valentindavid) [inline libfwupdplugin/fu-sbatlevel-section.c:57]: I suppose we just rely on `fu_firmware_parse_stream` doing `g_return_val_if_fail(G_IS_INPUT_STREAM(stream), FALSE)` to handle the rare case where `fu_partial_input_stream_new` returned `NULL`. I was wondering if it is normal we do not check the return of `fu_partial_input_stream_new`.

File: libfwupdplugin/fu-sbatlevel-section.c [hunk 3]
```diff
@@ -72,40 +72,69 @@ fu_sbatlevel_section_parse(FuFirmware *firmware,
...
-					    stream,
-					    header_offset + previous_addr,
-					    "previous",
-					    0,
-					    flags,
-					    error))
+	if (!fu_sbatlevel_section_add_entry(
+		firmware,
+		stream,
+		offset + fu_struct_sbat_level_section_header_get_previous(st),
```
Comment (valentindavid) [inline libfwupdplugin/fu-sbatlevel-section.c:83]: I am a bit confused here. I thought that the "previous" offset was relative to after the header (that is the 2 offsets in the beginning). But here w add it to `offset`, which I would think is the beginning of the section. Or am I missing something?

**Review by valentindavid — COMMENTED** [review]
I have a few comments, I want to be sure I understand the changes. All these offset calculation are pain to the brain. ;)

File: libfwupdplugin/fu-sbatlevel-section.c (hunk 3 above)
Comment (hughsie) [reply libfwupdplugin/fu-sbatlevel-section.c:83]: No, that's exactly why I asked you! I thought the offset was from the start. Do you have a known-good sbatlevel blob handy you could upload so I can try to recreate it please? Thanks!

File: libfwupdplugin/fu-sbatlevel-section.c (hunk 2 above)
Comment (hughsie) [reply libfwupdplugin/fu-sbatlevel-section.c:57]: Not sure I follow; why would `fu_partial_input_stream_new()` return with failure?

File: libfwupdplugin/fu-sbatlevel-section.c (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-sbatlevel-section.c:43]: I don't think it matters; I think it's probably preferable to pass in the size without the NULL to just match something like strlen().

File: libfwupdplugin/fu-sbatlevel-section.c (hunk 3 above)
Comment (hughsie) [reply libfwupdplugin/fu-sbatlevel-section.c:83]: Also, confusingly, the existing code uses FU_STRUCT_SBAT_LEVEL_SECTION_HEADER_OFFSET_PREVIOUS which is 0x4, i.e. neither the start or the end of the struct!

Comment (valentindavid) [reply libfwupdplugin/fu-sbatlevel-section.c:83]: Actually it is relative the offset table. Which comes after the version. So `FU_STRUCT_SBAT_LEVEL_SECTION_HEADER_OFFSET_PREVIOUS` was correct. It is only the version we have to skip.

See https://github.com/rhboot/shim/blob/5914984a1ffeab841f482c791426d7ca9935a5e6/sbat_var.S
The addresses  are  `sbat_var_automatic - sbat_var_payload_header` and `sbat_var_latest - sbat_var_payload_header`. And `sbat_var_payload_header` is just after the version.


Here is an example:
```
./usr/lib/shim/shimx64.efi.dualsigned:     file format pei-x86-64

Contents of section .sbatlevel:
 87000 00000000 08000000 22000000 73626174  ........"...sbat
 87010 2c312c32 30323230 35323430 300a6772  ,1,2022052400.gr
 87020 75622c32 0a007362 61742c31 2c323032  ub,2..sbat,1,202
 87030 32313131 3530300a 7368696d 2c320a67  2111500.shim,2.g
 87040 7275622c 330a00                      rub,3..         
```

Comment (valentindavid) [reply libfwupdplugin/fu-sbatlevel-section.c:83]: github does not let me upload a small binary file. You can base64 decode this if you want:

`AAAAAAgAAAAiAAAAc2JhdCwxLDIwMjIwNTI0MDAKZ3J1YiwyCgBzYmF0LDEsMjAyMjExMTUwMApzaGltLDIKZ3J1YiwzCgA=`

Comment (hughsie) [reply libfwupdplugin/fu-sbatlevel-section.c]: Super useful, thanks! PR updated -- does that look okay now?

Comment (valentindavid) [reply libfwupdplugin/fu-sbatlevel-section.c]: It does.

**Review by valentindavid — APPROVED** [review]

**Pull request merged**